client_internals.go

```diff
@@ -20,36 +20,174 @@
+
+var (
+	NullFrame  = NewFrameWith([]byte{}, IsNullFlag)
+	BeginFrame = NewFrameWith([]byte{}, BeginDataStructureFlag)
+	EndFrame   = NewFrameWith([]byte{}, EndDataStructureFlag)
+)
+
+func NewClientMessageForEncode() *ClientMessage {
+	return proto.NewClientMessageForEncode()
+}
+
+func NewFrame(content []byte) Frame {
+	return proto.NewFrame(content)
+}
+
+func NewFrameWith(content []byte, flags uint16) Frame {
+	return proto.NewFrameWith(content, flags)
+}
+
+func NewPair(key, value interface{}) Pair {
+	return Pair{Key: key, Value: value}
+}
+
+type InvokeOptions struct {
+	// Handler is the event handler for an invocation.
+	Handler ClientMessageHandler
+}
+
+// ClientInternal is an accessor for a Client.
 type ClientInternal struct {
 	client *Client
+	proxy  *proxy
 }
 
+// NewClientInternal creates the client internal accessor with the given client.
 func NewClientInternal(c *Client) *ClientInternal {
-	return &ClientInternal{client: c}
+	return &ClientInternal{
+		client: c,
+		proxy:  c.proxyManager.invocationProxy,
+	}
 }
 
-func (ci *ClientInternal) ConnectionManager() *cluster.ConnectionManager {
-	return ci.client.ic.ConnectionManager
+// Client returns the wrapped Client instance.
+func (ci *ClientInternal) Client() *Client {
+	return ci.client
 }
 
-func (ci *ClientInternal) DispatchService() *event.DispatchService {
-	return ci.client.ic.EventDispatcher
+// ClusterID returns the cluster ID.
+// It returns zero value of types.UUID{} if the cluster ID does not exist.
+func (ci *ClientInternal) ClusterID() types.UUID {
+	return ci.client.ic.ConnectionManager.ClusterID()
 }
 
-func (ci *ClientInternal) InvocationService() *invocation.Service {
-	return ci.client.ic.InvocationService
+// EncodeData serializes the given value and returns a Data value.
+func (ci *ClientInternal) EncodeData(obj interface{}) (Data, error) {
+	return ci.proxy.convertToData(obj)
 }
 
-func (ci *ClientInternal) InvocationHandler() invocation.Handler {
-	return ci.client.ic.InvocationHandler
+// DecodeData deserializes the given Data and returns a value.
+func (ci *ClientInternal) DecodeData(data Data) (interface{}, error) {
+	return ci.proxy.convertToObject(data)
 }
 
-func (ci *ClientInternal) ClusterID() types.UUID {
-	return ci.client.ic.ConnectionManager.ClusterID()
+// InvokeOnRandomTarget sends the given request to one of the members.
+// If opts.Handler is given, it is used as an event listener handler.
+func (ci *ClientInternal) InvokeOnRandomTarget(ctx context.Context, request *ClientMessage, opts *InvokeOptions) (*ClientMessage, error) {
+	var handler proto.ClientMessageHandler
+	if opts != nil {
+		handler = opts.Handler
+	}
+	return ci.proxy.invokeOnRandomTarget(ctx, request, handler)
+}
+
+// InvokeOnPartition sends the given request to the member which has the given partition ID.
+func (ci *ClientInternal) InvokeOnPartition(ctx context.Context, request *ClientMessage, partitionID int32, opts *InvokeOptions) (*ClientMessage, error) {
```

Comment:
`opts` is not used

Reply:
`opts` is there just for future expansion.